docs/api/StreamSend.md

```diff
@@ -21,15 +21,73 @@ QUIC_STATUS
 
 # Parameters
 
-**TODO**
+`Stream`
+
+The valid handle to an open stream object.
+
+`Buffers`
+
+An array of `QUIC_BUFFER` structs that each contain a pointer and length to app data to send on the stream. This may be `NULL` **only** if `BufferCount` is zero.
+
+`BufferCount`
+
+The number of `QUIC_BUFFER` structs in the `Buffers` array. This may be zero.
+
+`Flags`
+
+The set of flags the controls the behavior of `StreamSend`:
+
+Value | Meaning
+--- | ---
+**QUIC_SEND_FLAG_NONE**<br>0 | No special behavior. Data is not allowed in 0-RTT by default.
+**QUIC_SEND_FLAG_ALLOW_0_RTT**<br>1 | Indicates that the data is allowed to be sent in 0-RTT (if available). Makes no guarantee the data will be sent in 0-RTT. Additionally, even if 0-RTT keys are available the data may end up being sent in 1-RTT for multiple reasons.
+**QUIC_SEND_FLAG_START**<br>2 | Indicates that the stream should asynchronously start (equivalent to calling [StreamStart](StreamStart.md) with the `QUIC_STREAM_START_FLAG_ASYNC` flag).
+**QUIC_SEND_FLAG_FIN**<br>4 | Indicates the the stream send is the last or final data to be sent on the stream and should be gracefully shutdown (equivalent to calling [StreamShutdown](StreamShutdown.md) with the `QUIC_STREAM_SHUTDOWN_FLAG_GRACEFUL` flag).
+**QUIC_SEND_FLAG_DGRAM_PRIORITY**<br>8 | **Unused and ignored** for `StreamSend`
+**QUIC_SEND_FLAG_DELAY_SEND**<br>16 | Provides a hint to MsQuic to indicate the data does not need to be sent immediately, likely because more is soon to follow. MsQuic may still end up sending the data if something else triggers a flush of the connection-wide send queue.
```

Comment:
How does this work when I don't specify QUIC_STREAM_START_FLAG_IMMEDIATE to StreamStart?

If I use QUIC_SEND_FLAG_DELAY_SEND, then presumably nothing is guaranteed to be sent on the wire... but is it still true that the stream may end up getting created on the wire "if something else triggers a flush of the connection-wide send queue"?
If I don't use QUIC_SEND_FLAG_DELAY_SEND, then will this cause the stream to be started on the wire? I assume so, just checking.

Reply:
The `QUIC_STREAM_START_FLAG_IMMEDIATE` flag is purely there to support sending an empty stream frame to the peer; essentially triggering the PEER_STREAM_STARTED event but no actual data. That's it. It's completely independent of all these send flags.

If you make any send call that will trigger a stream frame to be sent out. The DELAY_SEND flag just says don't use this as the reason to flush the send queue. Something else might still come in and cause it to be flushed; but maybe not.

Does that answer your question?

Reply:
Also: 

When using QUIC_SEND_FLAG_DELAY_SEND, I think data will be sent anyway if you just have enough data on this stream to fill a packet, right?

The current wording is a bit confusing because it says "MsQuic may still end up sending the data if **something else** triggers a flush"... but in fact, this call alone can trigger than flush, right?

Reply:
If you have a completely idle connection with no other events (local or remote) and you make a StreamSend with the DELAY_SEND flag, then it's likely **nothing** will trigger your data to get flushed. The send by itself cannot cause the flush. If you later do something else, such as send/shutdown a different stream or the peer sends something that needs to be acknowledged or a keep alive timer fires, all of these will cause the connection-wide send queue to flush which will also end up sending the stream data.

Do you have any suggestions on how I could better improve this text to make this clear?

Reply:
> Do you have any suggestions

I made suggestions in the text section below -- that's probably the best place to explain this. For this text right here, maybe just deleted the last sentence -- "MsQuic may still end up sending the data if something else triggers a flush of the connection-wide send queue." -- and leave the full explanation for below.

Reply:
> The send by itself cannot cause the flush.

Okay, that's a little surprising to me. Maybe not a doc issue, but to make sure I understand:

Even if I send tons of data on the stream, it will all just get queued up?

Why not go ahead and send if you can fill a packet? Or if not just a single packet, then some threshold?

I'm wondering about HTTP3 scenarios with a large request body. Naively, I would implement this by just calling StreamSend with QUIC_SEND_FLAG_DELAY_SEND over and over again, until I run out of request body and call StreamSend with FIN but without DELAY_SEND. From the description you've given, this seems like it would not work very well...

Reply:
In very high performance scenarios we are able to batch many (40+) packets down the UDP stack in one send call. We could probably safely use that as the threshold, but anything less than that is up to interpretation. What if the app wants to queue 20 packet bursts? 10? 5? Any value less than the absolute max could be wrong.

The cost of the UDP send is still quite expensive, and the only way to get around that is batching. So I prefer to err on the side of improved batching than less. The app still has ultimate control here though... Don't use the flag on a send that meets some app-specific threshold.

Reply:
The result here will be that we'll need to implement our own threshold, right? I'd much rather that msquic determine the threshold than that we do. We will do a worse job.

Reply:
aside: are you using TransmitPackets to batch the packet sends? just curious...

Reply:
The API depends on the platform. Windows user mode uses `WSASendMsg`.

Reply:
For the threshold issue, feel free to open a separate issue.

Reply:
Will do.

Just so I understand: is there a limit to how much msquic will buffer? If so what is it?

Reply:
No limit. If you're using send buffering, yes there is a limit that we will buffer/copy locally. But you can always queue more than that, and it will just queue up. Without send buffering we just queue everything up.

Reply:
Yeah, I mean using send buffering, which we are currently doing. (We might change this in the future, but that's how it works today).

So how is someone supposed to use DELAY_SEND with send buffering? The user doesn't know how much data msquic has buffered. Any StreamSend call could result in the buffer being full, and I can't know this in advance. If I use DELAY_SEND, then I'll be waiting for a SEND_COMPLETE event that will never happen. Seems like the only option is to not use DELAY_SEND. Am I missing something?

Reply:
Send buffering and DELAY_SEND are a bit orthogonal IMO. The threshold of data you queue up to send in a batch is necessarily a function of how much you buffer locally. IMO, you should just use DELAY_SEND if you know there is more coming very soon and you'd really like it to be packaged up together. Otherwise, don't use it and let MsQuic do its thing.

Reply:
In the typical HTTP3 case there's always more data coming soon. We just don't know how much data there will be.

Reply:
My recommendation: for a given HTTP payload that requires qpack + data to be sent on their respective streams, queue the qpack data with DELAY_SEND and the other data without it.

Reply:
That seems like it will hurt perf, no?

Reply:
Not necessarily. The only way to make an accurate conclusion is to measure.

Reply:
Fair point.

Reply:
That said, I still don't quite see how I can reliably use DELAY_SEND with send buffering even in the simple QPACK case you suggest.

Currently, we will never post more than one send at a time to msquic. We will always wait until the previous send completes before posting another send. With send buffering, the sends will complete reasonably promptly until the buffer is full, and then will not. In general this is fine -- the send will complete eventually (or fail, or we will timeout, etc).

But with DELAY_SEND, it seems like it's always possible (though possibly very unlikely) that the send will *never* complete *until* we do another send without DELAY_SEND. 

I think that makes DELAY_SEND unusable for us as long as we are using send buffering.

Reply:
Personally, I'd recommend **not** using send buffering purely because there is a significant throughput penalty (8106766 vs 7082789). This seems to just be yet more reason to move off it.

Reply:
Yes, that's the conclusion I come to as well -- we need to move away from using msquic send buffering at some point. I think it's still a fine starting point, and we might ship .NET 6 with it, but we can't use it with DELAY_SEND, which makes it less useful. There are other reasons for us to move away from msquic send buffering anyway, like supporting synchronous completions.

That said, this will mean that we implement _our own_ send buffering, because that's how our model works. This should work better for our purposes than msquic send buffering, but it will still have the overhead of copying into a send buffer and managing the buffer etc. So this will very likely still have some penalty vs msquic style unbuffered send.

In the long term it's possible that we could support an "unbuffered send" API in managed code, but the complexity of this for managed code users is pretty high, so I am skeptical that we will ever actually do it. We'll need to explore what the potential perf win is and how we could expose this to users in the best way.

Reply:
BTW, the model of doing one send at a time is likely going to be a significant perf problem even if we fixed the delay send issue (which I'm sure we could come up with a heuristic to immediately flush).

Reply:
How significant, do you expect?